ARROW-6242: [C++][Dataset] Implement Dataset, Scanner and ScannerBuilder

fsaintjacques opened the pull request:
This adds the first Scanner implementation that binds all classes. This PR also patches previous requests regarding testing functions/classes.

kszucs commented:
@ursabot build

kszucs commented:
@ursabot build

pprudhvi commented on cpp/src/arrow/dataset/dataset_test.cc:
```diff
@@ -18,36 +18,37 @@
 #include "arrow/dataset/dataset.h"
 
 #include "arrow/dataset/test_util.h"
```
this requires parquet libraries to be built in order to build arrow.

The error when building arrow with -DARROW_PARQUET=OFF is
```
'parquet/parquet_version.h' file not found
In file included from cpp/src/arrow/dataset/test_util.h:31:
In file included from cpp/src/parquet/arrow/writer.h:25:
```

pitrou commented:
There's an ASAN failure on Travis-CI. Also, can you rebase?

fsaintjacques commented:
I'm looking into the ASAN failure, I'll rebase.

fsaintjacques commented:
Turns out, clang-8 couldn't find, but using clang-7 worked.

fsaintjacques commented:
ASAN fixed https://travis-ci.org/fsaintjacques/arrow/jobs/578347359

pitrou commented on cpp/src/arrow/util/iterator.h:
```diff
@@ -62,14 +62,48 @@ class Iterator {
 };
 
+template <typename Fn, typename T>
+class FunctionIterator : public Iterator<T> {
```
Hmm... This code is already in master, right? Did you mess a rebase?

pitrou replied:
Oh, I see, the diff is messy.

fsaintjacques commented:
Finally passing... https://travis-ci.org/fsaintjacques/arrow/builds/578395773

pitrou commented on cpp/src/arrow/dataset/dataset_test.cc:
```diff
@@ -58,5 +59,33 @@ TEST_F(TestSimpleDataSource, GetFragments) {
+      std::make_shared<SimpleDataSource>(fragments),
+  };
+
+  const int64_t total_batches = sources.size() * kNumberBatches * kNumberBatches;
```
Hmm... shouldn't that be `sources.size() * kNumberBatches * kNumberFragments`? We have 2 sources, each with kNumberFragments, each with kNumberBatches.

pitrou commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -76,16 +76,55 @@ class ARROW_DS_EXPORT SimpleScanTask : public ScanTask {
+///  - DataSource::GetFragments
+///  - DataFragment::Scan
+///
+/// Thus, this iterator is not suited for consumption of sources/fragments
```
Do you have an idea how a non-trivial Scanner would work in order to hide latencies?

fsaintjacques replied:
- DataSource can cache the list of directories/files, or force an evaluation at creation time, instead of GetFragments.
- DataFragment::Scan is the hardest. For S3 files, you hope that you can "peek" the metadata and make the partitioning in ScanTask from there, e.g. with parquet, you can in theory only read the metadata (via 2 tiny ReadAt) an dispatch a ScanTask per RowGroup. For CSV, you could have companion metadata file that gives offsets of chunks in bytes. The trick is to defer the high latency into the ScanTask.

As for the scanner itself, you can "emulate" this by giving the Scanner internal threads that consume from DataSources and push to a queue, and have ScanTask that hides a future that reads from said queue or yield an empty iterator if everything was consumed. In practice, I will expect that we focus on the pushing the latency in the DataFragments by deferring downloads.

As for your interest, we'll need to refactor the CSV parser to provide an interface where we expose the Chunks (offset of bytes containing full lines?) and bind one ScanTask per Chunk, assuming that type inference is disabled.

pitrou commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -76,16 +76,55 @@ class ARROW_DS_EXPORT SimpleScanTask : public ScanTask {
+/// where the previous methods can block for a long time, e.g. if fetching a
+/// DataFragment from cloud storage, or a DataFragment must be parsed before
+/// returning a ScanTaskIterator.
+class SimpleScanner : public Scanner {
```
ARROW_DS_EXPORT, no?

pitrou commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -104,7 +143,7 @@ class ARROW_DS_EXPORT ScannerBuilder {
   ScannerBuilder* IncludePartitionKeys(bool include = true);
 
```
Not in this PR, but the docstring for `IncludePartitionKeys` is incomplete ("otherwise"... what?).

pitrou commented on cpp/src/arrow/dataset/scanner_test.cc:
```diff
@@ -0,0 +1,56 @@
+      std::make_shared<SimpleDataSource>(fragments),
+  };
+
+  const int64_t total_batches = sources.size() * kNumberBatches * kNumberBatches;
```
Same question as in the other test.

pitrou replied:
Also, if there's a genuine error here, it's curious the test doesn't catch it.

pitrou commented on cpp/src/arrow/dataset/test_util.h:
```diff
@@ -60,34 +60,31 @@ class DatasetFixtureMixin : public ::testing::Test {
-};
+  /// \brief Ensure that record batches found in reader are equals to the
+  /// record batches yielded by the data fragment.
+  void AssertScanTaskEquals(RecordBatchReader* expected, ScanTask* task) {
```
At the end, this should also check that `expected` is exhausted (otherwise, the sequences were not the same length).

pitrou commented on cpp/src/arrow/dataset/test_util.h:
```diff
@@ -98,6 +95,35 @@ class TestDataSourceMixin : public TestDataFragmentMixin {
+    ARROW_EXPECT_OK(it->Visit([&](std::unique_ptr<ScanTask> task) -> Status {
+      AssertScanTaskEquals(expected, task.get());
+      return Status::OK();
+    }));
```
Also needs to check that `expected` is exhausted at the end.

pitrou commented on cpp/src/arrow/dataset/test_util.h:
```diff
@@ -98,6 +95,35 @@ class TestDataSourceMixin : public TestDataFragmentMixin {
+    std::unique_ptr<Scanner> scanner;
+    ASSERT_OK(builder->Finish(&scanner));
+
+    auto it = scanner->Scan();
```
This variables isn't used.

pitrou commented on cpp/src/arrow/record_batch.h:
```diff
@@ -189,4 +189,15 @@ class ARROW_EXPORT RecordBatchReader
   Status ReadAll(std::shared_ptr<Table>* table);
 };
 
+/// \brief Creates a RecordBatchReader from a vector  of RecordBatch.
```
Generally we use infinitives in docstrings, so "Create" not "Creates".

pitrou commented on cpp/src/arrow/record_batch.cc:
```diff
@@ -269,4 +269,41 @@ Status RecordBatchReader::ReadAll(std::shared_ptr<Table>* table) {
+                             std::shared_ptr<Schema> schema,
+                             std::shared_ptr<RecordBatchReader>* out) {
+  if (schema == nullptr) {
+    if (batches.size() == 0 || batches[0] == nullptr) {
```
Hmm... is a nullptr expected here?

pitrou commented on cpp/src/arrow/record_batch.cc:
```diff
@@ -269,4 +269,41 @@ Status RecordBatchReader::ReadAll(std::shared_ptr<Table>* table) {
 
+class SimpleRecordBatchReader : public RecordBatchReader {
+ public:
+  SimpleRecordBatchReader(std::unique_ptr<Iterator<std::shared_ptr<RecordBatch>>> it,
```
This constructor isn't used, is it?

pitrou commented on cpp/src/arrow/util/iterator.h:
```diff
@@ -106,44 +143,100 @@ class MapIterator : public Iterator<O> {
+        typename std::remove_pointer<internal::call_traits::argument_type<0, Fn>>::type,
+    typename O =
+        typename std::remove_pointer<internal::call_traits::argument_type<1, Fn>>::typ>
+class MaybeMapIterator : public Iterator<O> {
```
Is there a point in having both `MapIterator` and `MaybeMapIterator`? AFAICS, these are not meant to be used for yielding very small datum, so shaving a couple CPU cycles doesn't seem important here.

pitrou commented on cpp/src/arrow/util/iterator.h:
```diff
@@ -106,44 +143,100 @@ class MapIterator : public Iterator<O> {
+      return Next(out);
+    }
+
+    // Pop from current_ and lookout for depletion.
```
s/current_/child_/

pitrou commented on cpp/src/arrow/util/iterator.h:
```diff
@@ -106,44 +143,100 @@ class MapIterator : public Iterator<O> {
+
+    // Pop from current_ and lookout for depletion.
+    ARROW_RETURN_NOT_OK(child_->Next(out));
+    if (out == NULLPTR) {
```
Hmm... shouldn't that be `*out == NULLPTR`? Is this facility tested?

kszucs commented:
@ursabot build

fsaintjacques commented:
@pitrou, I added iterator_test.cc that confirmed the bug (and fix). I also added the check for consumption in dataset test utils, which confirmed the wrong count bug (and fix). Thanks for the thorough review, it caught 3 intertwined bugs!

pitrou commented:
@fsaintjacques Can you rebase? There's a conflict now.
